src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -99,6 +104,166 @@ using JoystickBlacklist = std::vector<JoystickBlacklistEntry>;
+    if (data.bstrDeviceID == nullptr)
+        return false;
+
+    // Connect to WMI
+    hr = data.pIWbemLocator->ConnectServer(data.bstrNamespace, nullptr, nullptr, nullptr, 0, nullptr, nullptr, &data.pIWbemServices);
+    if (FAILED(hr) || data.pIWbemServices == nullptr)
+        return false;
+
+    // Switch security level to IMPERSONATE.
+    hr = CoSetProxyBlanket(data.pIWbemServices,
+                           RPC_C_AUTHN_WINNT,
+                           RPC_C_AUTHZ_NONE,
+                           nullptr,
+                           RPC_C_AUTHN_LEVEL_CALL,
+                           RPC_C_IMP_LEVEL_IMPERSONATE,
+                           nullptr,
+                           EOAC_NONE);
+    if (FAILED(hr))
+        return false;
+
+    hr = data.pIWbemServices->CreateInstanceEnum(data.bstrClassName, 0, nullptr, &data.pEnumDevices);
+    if (FAILED(hr) || data.pEnumDevices == nullptr)
+        return false;
+
+    // Loop over all devices
+    for (;;)
+    {
+        ULONG uReturned = 0;
+        hr = data.pEnumDevices->Next(10'000, static_cast<ULONG>(data.pDevices.size()), data.pDevices.data(), &uReturned);
+        if (FAILED(hr))
+            return false;
+
+        if (uReturned == 0)
+            break;
+
+        for (std::size_t iDevice = 0; iDevice < uReturned; ++iDevice)
+        {
+            // For each device, get its device ID
+            hr = data.pDevices[iDevice]->Get(data.bstrDeviceID, 0L, &data.var, nullptr, nullptr);
+            if (SUCCEEDED(hr) && data.var.vt == VT_BSTR && data.var.bstrVal != nullptr)
+            {
+                // Check if the device ID contains "IG_".  If it does, then it's an XInput device
+                // This information cannot be found from DirectInput
+                if (std::wcsstr(data.var.bstrVal, L"IG_"))
+                {
+                    // If it does, then get the VID/PID from var.bstrVal
+                    DWORD  dwPid  = 0;
+                    DWORD  dwVid  = 0;
+                    WCHAR* strVid = std::wcsstr(data.var.bstrVal, L"VID_");
+                    if (strVid && swscanf_s(strVid, L"VID_%4X", &dwVid) != 1)
+                        dwVid = 0;
+                    WCHAR* strPid = std::wcsstr(data.var.bstrVal, L"PID_");
+                    if (strPid && swscanf_s(strPid, L"PID_%4X", &dwPid) != 1)
+                        dwPid = 0;
+
+                    // Compare the VID/PID to the DInput device
+                    const auto dwVidPid = static_cast<DWORD>(MAKELONG(dwVid, dwPid));
+                    if (dwVidPid == pGuidProductFromDirectInput->Data1)
+                        return true;
+                }
+            }
+
+            VariantClear(&data.var);
+            auto* device = data.pDevices[iDevice];
+            if (device)
+            {
+                device->Release();
+                // important to prevent a double-free in cleanup data destructor
+                device = nullptr;
```

Comment:
This is incorrect, it should zero out `data.pDevices[iDevice]`.
Now you are just clearing your local variable `device`, but that does not update the actual pDevices.

Reply:
Calling a release function then setting a pointer to null makes me think this could be a good place to use `std::unique_ptr` with a custom deleter.